imap_processing/ultra/l1c/ultra_l1c_pset_bins.py

```diff
@@ -44,11 +46,131 @@ def build_spatial_bins(spacing: float = 0.5) -> tuple[np.ndarray, np.ndarray]:
         Array of azimuth bin boundary values.
     el_bin_edges : np.ndarray
         Array of elevation bin boundary values.
+    az_bin_midpoints : np.ndarray
+        Array of azimuth bin midpoint values.
+    el_bin_midpoints : np.ndarray
+        Array of elevation bin midpoint values.
     """
     # Azimuth bins from 0 to 360 degrees.
     az_bin_edges = np.arange(0, 360 + spacing, spacing)
+    az_bin_midpoints = az_bin_edges[:-1] + spacing / 2  # Midpoints between edges
 
     # Elevation bins from -90 to 90 degrees.
     el_bin_edges = np.arange(-90, 90 + spacing, spacing)
+    el_bin_midpoints = el_bin_edges[:-1] + spacing / 2  # Midpoints between edges
 
-    return az_bin_edges, el_bin_edges
+    return az_bin_edges, el_bin_edges, az_bin_midpoints, el_bin_midpoints
+
+
+def cartesian_to_spherical(
+    vx: np.ndarray, vy: np.ndarray, vz: np.ndarray
+) -> tuple[np.ndarray, np.ndarray]:
+    """
+    Convert cartesian coordinates to spherical coordinates.
+
+    Parameters
+    ----------
+    vx : np.ndarray
+        The x-components of the velocity vector.
+    vy : np.ndarray
+        The y-components of the velocity vector.
+    vz : np.ndarray
+        The z-components of the velocity vector.
+
+    Returns
+    -------
+    az : np.ndarray
+        The azimuth angles in degrees.
+    el : np.ndarray
+        The elevation angles in degrees.
+    """
+    # Magnitude of the velocity vector
+    magnitude_v = np.sqrt(vx**2 + vy**2 + vz**2)
+
+    vhat_x = -vx / magnitude_v
+    vhat_y = -vy / magnitude_v
+    vhat_z = -vz / magnitude_v
+
+    # Convert from cartesian to spherical coordinates (azimuth, elevation)
+    # Radius (magnitude)
+    r = np.sqrt(vhat_x**2 + vhat_y**2 + vhat_z**2)
+
+    # Elevation angle (angle from the z-axis, range: [-pi/2, pi/2])
+    el = np.arcsin(vhat_z / r)
+
+    # Azimuth angle (angle in the xy-plane, range: [0, 2*pi])
+    az = np.arctan2(vhat_y, vhat_x)
+
+    # Ensure azimuth is from 0 to 2PI
+    az = az % (2 * np.pi)
+
+    return az, el
+
+
+def bin_space(
+    vx: np.ndarray, vy: np.ndarray, vz: np.ndarray
+) -> tuple[np.ndarray, np.ndarray]:
+    """
+    Bin the particle.
+
+    Parameters
+    ----------
+    vx : np.ndarray
+        The x-components of the velocity vector.
+    vy : np.ndarray
+        The y-components of the velocity vector.
+    vz : np.ndarray
+        The z-components of the velocity vector.
+
+    Returns
+    -------
+    az_midpoint : np.ndarray
```

Comment:
I made some changes, but the reason I can't do a 2d histogram is that it would result in 259200 bins since the Ultra has 0.5 x 0.5 degree bins. I want my xarray to be like this:

coordinate: bin_id
variable: az_midpoint
variable: el_midpoint
variable: counts